### PR TITLE
SE-0075 Implement canImport

### PR DESCRIPTION
**Pull request by CodaFi**
SE-0075 requests that we offer a build-time check for the importability of modules.  This check, however, is particularly expensive so we can no longer resolve every condition statically.  Conditions that are not essential to the health of the parser are left unevaluated and deferred to a new phase post-parse called "Condition Resolution".  Condition Resolution will also skip functions that have been deferred to keep Code Completion speedy; resolution can be done on-demand.

Resolves [SR-1560](https://bugs.swift.org/browse/SR-1560).

### CODE REVIEW

**Comment by CodaFi** [comment]
@jrose-apple for Parse, @DougGregor for Sema. 

Come On Down

📯 📯 📯

File: lib/Sema/NameBinding.cpp [hunk 1]
```diff
@@ -297,6 +298,470 @@ static void insertPrecedenceGroupDecl(NameBinder &binder, SourceFile &SF,
...
+      ASTContext &Context = Binder.Context;
+      SmallVector<ASTNode, 3> Elements;
+      for (auto &Element : BS->getElements()) {
+        if (Stmt *stmt = Element.dyn_cast<Stmt *>()) {
+          if (IfConfigStmt *Config = dyn_cast<IfConfigStmt>(stmt)) {
+            for (auto &clause : Config->getClauses()) {
+              // Evaluate conditions until we find the active clause.
+              if (!clause.Cond || evaluateCondition(clause.Cond)) {
+                for (auto &e : clause.Elements) {
+                  if (auto innerStmt = e.dyn_cast<Stmt *>()) {
```
Comment (CodaFi) [inline lib/Sema/NameBinding.cpp]: I will try to un-pyramid-of-doom this mess.

File: lib/Sema/NameBinding.cpp [hunk 2]
```diff
@@ -297,6 +298,551 @@ static void insertPrecedenceGroupDecl(NameBinder &binder, SourceFile &SF,
...
+                                                 versionStartLoc,
+                                                 &Binder.Context.Diags);
+          trans.abort();
+
+          assert(versionRequirement.hasValue());
+          auto thisVersion = Context.LangOpts.EffectiveLanguageVersion;
+          assert(prefix->getName().getBaseName().str().equals(">="));
+
+          auto VersionNewEnough = thisVersion >= versionRequirement.getValue();
+          return VersionNewEnough;
```
Comment (jrose-apple) [inline lib/Sema/NameBinding.cpp]: I don't think this is correct. `#if swift(…)` (and `#if _compiler_version(…)`) really do need to be handled at parse time, because the parser *cannot handle* the code in the inactive case. Turning off diagnostics isn't sufficient.

File: lib/Sema/NameBinding.cpp [hunk 3]
```diff
@@ -297,6 +298,551 @@ static void insertPrecedenceGroupDecl(NameBinder &binder, SourceFile &SF,
...
+        } else {
+          auto UDRE = cast<UnresolvedDeclRefExpr>(PE->getSubExpr());
+
+          // The sub expression should be an UnresolvedDeclRefExpr (we won't
+          // tolerate extra parens).
+          auto argument = UDRE->getName().getBaseName().str();
+          // FIXME: Perform the replacement macOS -> OSX in Parse if possible.
```
Comment (jrose-apple) [inline lib/Sema/NameBinding.cpp]: Also, "finish making 'macOS' the canonical name". \*sigh\* Unfinished work from me.

(not really part of the review, just complaining out loud)

File: lib/Sema/NameBinding.cpp [hunk 4]
```diff
@@ -297,6 +298,551 @@ static void insertPrecedenceGroupDecl(NameBinder &binder, SourceFile &SF,
...
+                                        IfConfigDecl *Cond) {
+      if (Cond->isResolved()) return;
+      for (auto &clause : Cond->getClauses()) {
+        // Evaluate conditions until we find the active clause.
+        if (!clause.Cond || evaluateCondition(clause.Cond)) {
+          for (auto &member : clause.Members) {
+            if (auto innerConfig = dyn_cast<IfConfigDecl>(member)) {
+              resolveClausesAndInsertMembers(Nom, innerConfig);
+            } else {
+              Nom->addMember(member);
```
Comment (jrose-apple) [inline lib/Sema/NameBinding.cpp]: Hm. How much do you think it matters that this list used to be sorted?

File: lib/Sema/NameBinding.cpp [hunk 5]
```diff
@@ -297,6 +298,551 @@ static void insertPrecedenceGroupDecl(NameBinder &binder, SourceFile &SF,
...
+
+      llvm_unreachable("Unhandled condition kind?");
+    }
+
+    void resolveClausesAndInsertMembers(IterableDeclContext *Nom,
+                                        IfConfigDecl *Cond) {
+      if (Cond->isResolved()) return;
+      for (auto &clause : Cond->getClauses()) {
+        // Evaluate conditions until we find the active clause.
+        if (!clause.Cond || evaluateCondition(clause.Cond)) {
```
Comment (jrose-apple) [inline lib/Sema/NameBinding.cpp]: Early exit?

File: lib/Sema/NameBinding.cpp [hunk 6]
```diff
@@ -297,6 +298,551 @@ static void insertPrecedenceGroupDecl(NameBinder &binder, SourceFile &SF,
...
+          }
+          break;
+        }
+      }
+      Cond->setResolved();
+    }
+
+    Stmt *transformStmt(Stmt *S) {
+      switch (S->getKind()) {
+        case StmtKind::Brace:
```
Comment (jrose-apple) [inline lib/Sema/NameBinding.cpp]: Nitpick: LLVM style doesn't indent `case`.

File: lib/Sema/NameBinding.cpp [hunk 7]
```diff
@@ -297,6 +298,551 @@ static void insertPrecedenceGroupDecl(NameBinder &binder, SourceFile &SF,
...
+        case StmtKind::DoCatch:
+          return transformDoCatchStmt(cast<DoCatchStmt>(S));
+        default:
+          return S;
+      }
+    }
+
+    // transform*() return their input if it's unmodified,
+    // or a modified copy of their input otherwise.
+    IfStmt *transformIfStmt(IfStmt *IS) {
```
Comment (jrose-apple) [inline lib/Sema/NameBinding.cpp]: Do we really not have a general StmtTransformer yet? Okay.

File: lib/Sema/NameBinding.cpp [hunk 8]
```diff
@@ -297,6 +298,551 @@ static void insertPrecedenceGroupDecl(NameBinder &binder, SourceFile &SF,
...
+      if (Stmt *TS = IS->getThenStmt()) {
+        Stmt *NTS = transformStmt(TS);
+        if (NTS != TS) {
+          IS->setThenStmt(NTS);
```
Comment (jrose-apple) [inline lib/Sema/NameBinding.cpp]: This isn't a copy.

File: lib/Sema/NameBinding.cpp [hunk 9]
```diff
@@ -297,6 +298,551 @@ static void insertPrecedenceGroupDecl(NameBinder &binder, SourceFile &SF,
...
+        for (auto &clause : config->getClauses()) {
+          // Evaluate conditions until we find the active clause.
+          if (clause.Cond && !evaluateCondition(clause.Cond)) {
+            continue;
+          }
+
+          // Now that we have the active clause, look into its elements for
+          // further statements to transform.
+          for (auto &clauseElt : clause.Elements) {
+            auto innerStmt = clauseElt.dyn_cast<Stmt *>();
```
Comment (jrose-apple) [inline lib/Sema/NameBinding.cpp]: What happened to looking into closures?

File: lib/Sema/NameBinding.cpp [hunk 10]
```diff
@@ -297,6 +298,551 @@ static void insertPrecedenceGroupDecl(NameBinder &binder, SourceFile &SF,
...
+          break;
+        }
+      }
+
+      return swift::BraceStmt::create(Context, BS->getLBraceLoc(),
+                                      Context.AllocateCopy(Elements),
+                                      BS->getRBraceLoc());
+    }
+
+    BraceStmt *resolveClausesAndExtractTLCDs(BraceStmt *BS) {
```
Comment (jrose-apple) [inline lib/Sema/NameBinding.cpp]: A *lot* of this is the same as the normal `transformBraceStmt`. Think there's a way to factor out the common parts? Maybe with `llvm::function_ref` for callbacks?

File: lib/Sema/NameBinding.cpp [hunk 11]
```diff
@@ -297,6 +298,551 @@ static void insertPrecedenceGroupDecl(NameBinder &binder, SourceFile &SF,
...
+                  }
+                }
+                ExtraTLCDs.push_back(TLCD);
+              } else {
+                // For newly-inserted members of aggregates and extensions we
+                // need to warn if they're marked '@objc'.
+                if (auto *IDC = dyn_cast<IterableDeclContext>(TLD)) {
+                  for (auto member : IDC->getMembers()) {
+                    for (auto Attr : member->getAttrs()) {
+                      if (isa<ObjCAttr>(Attr) || isa<DynamicAttr>(Attr))
```
Comment (jrose-apple) [inline lib/Sema/NameBinding.cpp]: What about
- `@objc` on top-level decls?
- `@objc` on members of local decls?

File: lib/Sema/NameBinding.cpp [hunk 12]
```diff
@@ -297,6 +298,551 @@ static void insertPrecedenceGroupDecl(NameBinder &binder, SourceFile &SF,
...
+
+        if (BraceStmt *Body = TLCD->getBody()) {
+          BraceStmt *NewBody = resolveClausesAndExtractTLCDs(Body);
+          if (NewBody != Body) {
+            TLCD->setBody(NewBody);
+          }
+        }
+      } else if (AbstractFunctionDecl *FD = dyn_cast<AbstractFunctionDecl>(D)) {
+        // For functions, transform the statements in the body.
+        // FIXME: This can be deferred until type checking.
```
Comment (jrose-apple) [inline lib/Sema/NameBinding.cpp]: Yeah, I think we do need to figure out how to do this. Code completion doesn't parse function bodies unless it needs to, but this will probably fault them all in eagerly.

File: lib/Sema/NameBinding.cpp [hunk 13]
```diff
@@ -319,6 +865,32 @@ void swift::performNameBinding(SourceFile &SF, unsigned StartElem) {
...
+    if (auto TLD = dyn_cast<TopLevelCodeDecl>(d)) {
+      // When parsing library files, skip empty top level code declarations.
+      // These are artifacts of the condition clause transform for top level
+      // declarations.
+      if (!SF.isScriptMode() && TLD->getBody()->getElements().empty()) continue;
+    }
+    SmallVector<Decl *, 8> scratch;
+    auto processedD = resolveConditionalClauses(d, Binder, scratch);
+    assert(scratch.empty());
+    SF.Decls.push_back(processedD);
```
Comment (jrose-apple) [inline lib/Sema/NameBinding.cpp]: TopLevelCodeDecls *definitely* need to stay in source order.

File: test/ClangImporter/enum-error.swift [hunk 14]
```diff
@@ -91,6 +91,7 @@ func testError() {
   } catch is TestError {
   } catch {}
 
+  print(testErrorNSError)
```
Comment (jrose-apple) [inline test/ClangImporter/enum-error.swift]: Unrelated change?

File: test/IDE/print_ast_tc_decls.swift [hunk 15]
```diff
@@ -1330,14 +1330,15 @@ protocol ProtocolToExtend {
...
+// Should be evaluated before type checking.
 #if true
 #elseif false
 #else
 #endif
-// PASS_PRINT_AST: #if
-// PASS_PRINT_AST: #elseif
-// PASS_PRINT_AST: #else
-// PASS_PRINT_AST: #endif
+// PASS_PRINT_AST-NOT: #if
```
Comment (jrose-apple) [inline test/IDE/print_ast_tc_decls.swift]: I don't think this is correct either. It's important to have the document structure in place for SourceKit and other tools.

File: test/decl/var/usage.swift [hunk 16]
```diff
@@ -167,8 +167,8 @@ func testForceValueExpr() {
 
 // <rdar://problem/20894455> "variable was never mutated" diagnostic does not take #if into account
 func testBuildConfigs() {
-  let abc = 42    // no warning.
-  var mut = 18    // no warning.
+  let abc = 42    // expected-warning{{initialization of immutable value 'abc' was never used; consider replacing with assignment to '_' or removing it}}.
+  var mut = 18    // expected-warning{{initialization of variable 'mut' was never used; consider replacing with assignment to '_' or removing it}}.
```
Comment (jrose-apple) [inline test/decl/var/usage.swift]: Just to make it clear, this is a regression.

**Review by jrose-apple — COMMENTED** [review]
Preliminary comments on your WIP.

File: include/swift/AST/Decl.h [hunk 17]
```diff
@@ -1895,32 +1890,24 @@ class IfConfigDecl : public Decl {
   /// An array of clauses controlling each of the #if/#elseif/#else conditions.
   /// The array is ASTContext allocated.
   ArrayRef<IfConfigDeclClause> Clauses;
+  SourceLoc StartLoc;
```
Comment (CodaFi) [inline include/swift/AST/Decl.h]: At one point I had caused destructive updates to the clause array, but that was ridiculous and I reverted it.  Going to revert this part too.

File: lib/Parse/ParseDecl.cpp [hunk 18]
```diff
@@ -5092,6 +5089,9 @@ ParserResult<ClassDecl> Parser::parseDeclClass(SourceLoc ClassLoc,
     Scope S(this, ScopeKind::ClassBody);
     ParseDeclOptions Options(PD_HasContainerType | PD_AllowDestructor |
                              PD_InClass);
+    if (Flags & PD_InIfConfig) {
+      Options |= PD_InIfConfig;
```
Comment (CodaFi) [inline lib/Parse/ParseDecl.cpp]: There was no flag for this before.  It used to be a function of "scope", but I felt it was inappropriate for scope to ask the question "is this a condition" let alone "is this an inactive condition" like it did before.

File: lib/Sema/NameBinding.cpp (hunk 2 above)
Comment (CodaFi) [inline lib/Sema/NameBinding.cpp]: I was afraid of this.  It might be useful to have the new stripped-down `classify` function set a trinary value indicating a clause was evaluated then and there and assert down here that we've already handled this.

File: lib/Sema/NameBinding.cpp (hunk 5 above)
Comment (CodaFi) [inline lib/Sema/NameBinding.cpp]: 👍

File: lib/Sema/NameBinding.cpp (hunk 9 above)
Comment (CodaFi) [inline lib/Sema/NameBinding.cpp]: Incoming.

File: test/IDE/print_ast_tc_decls.swift (hunk 15 above)
Comment (CodaFi) [inline test/IDE/print_ast_tc_decls.swift]: Hmph, re-inserting `IfConfigStmt`s should do the trick, then.

File: test/decl/var/usage.swift (hunk 16 above)
Comment (CodaFi) [inline test/decl/var/usage.swift]: The usage clause is inactive, why shouldn't we warn here exactly?

File: test/ClangImporter/enum-error.swift (hunk 14 above)
Comment (CodaFi) [inline test/ClangImporter/enum-error.swift]: Related to the `test/decl/var/usage.swift` regression.

File: lib/Sema/NameBinding.cpp (hunk 10 above)
Comment (CodaFi) [inline lib/Sema/NameBinding.cpp]: Precisely what I was thinking.  There's a huge amount of overlap here with playground logger that can be resolved with a general `StmtTransformer` too.

File: test/decl/var/usage.swift (hunk 16 above)
Comment (jrose-apple) [inline test/decl/var/usage.swift]: Because it might be active under another build configuration, and there's no way to silence it otherwise.

**Comment by CodaFi** [comment]
Let's see what happens.

@swift-ci please smoke test.

**Comment by CodaFi** [comment]
@swift-ci please smoke test.

**Comment by CodaFi** [comment]
Let's get rid of some of the noise

@swift-ci please smoke test.

**Comment by CodaFi** [comment]
@swift-ci please smoke test.

**Comment by CodaFi** [comment]
@swift-ci please smoke test Linux platform.

**Comment by CodaFi** [comment]
@swift-ci please smoke test.

**Comment by CodaFi** [comment]
@swift-ci please smoke test.

**Comment by CodaFi** [comment]
@swift-ci please smoke test.

**Comment by CodaFi** [comment]
@swift-ci please smoke test.

**Comment by CodaFi** [comment]
@swift-ci please smoke test.

**Comment by CodaFi** [comment]
@swift-ci please smoke test.

**Comment by CodaFi** [comment]
70th time's the charm.

@swift-ci please smoke test.

**Comment by CodaFi** [comment]
@swift-ci please test.

**Comment by CodaFi** [comment]
@jrose-apple please re-review now that the tests pass.

**Comment by CodaFi** [comment]
The LLDB failure is unrelated, the new crasher being non-deterministic or possibly platform-dependent is.  @practicalswift was this meant to be non-deterministic?

**Comment by CodaFi** [comment]
@swift-ci please smoke test.

**Comment by CodaFi** [comment]
@swift-ci please test.

**Comment by CodaFi** [comment]
@swift-ci please test.

**Comment by CodaFi** [comment]
@swift-ci please smoke test.

**Comment by CodaFi** [comment]
@swift-ci please smoke test.

**Comment by CodaFi** [comment]
@swift-ci please test Linux platform.

**Comment by CodaFi** [comment]
@jrose-apple ping for review.

File: lib/Sema/PlaygroundTransform.cpp [hunk 19]
```diff
@@ -55,14 +53,14 @@ template <class E> class Added {
   E &operator->() { return Contents; }
 };
 
-class Instrumenter {
+class Instrumenter : StmtTransformer {
```
Comment (jrose-apple) [inline lib/Sema/PlaygroundTransform.cpp]: Let's leave PlaygroundTransform out of this for now.

**Review by jrose-apple — COMMENTED** [review]
I don't think the post-processor strategy is going to work well with lazily-parsed bodies, and I don't like that it requires an additional pointer in every AbstractFunctionDecl. Maybe this should just be a new body-stage between Parsed and TypeChecked that's handled on demand, though I'm not sure if SourceKit will want to resolve as many `#if`s as possible when asking for something like a syntax map.

(This is not a detail-level review just yet.)

**Comment by jrose-apple** [comment]
Thanks for taking this on, though. Implementing this proposal is a tough nut to crack, and I think you've got the best approach. I wonder if we'll have to think about optimization, though; this is more N^2 work being added to the compiler, and we should really really try to minimize that. cc @graydon

File: lib/Sema/PlaygroundTransform.cpp (hunk 19 above)
Comment (CodaFi) [inline lib/Sema/PlaygroundTransform.cpp]: Can do.  Shall I remove StmtTransformer as well?

File: include/swift/AST/Decl.h [hunk 20]
```diff
@@ -1893,13 +1892,13 @@ class IfConfigDecl : public Decl {
   ArrayRef<IfConfigDeclClause> Clauses;
   SourceLoc EndLoc;
   bool HadMissingEnd;
+  bool HasBeenResolved = false;
```
Comment (DougGregor) [inline include/swift/AST/Decl.h]: Oh, the extra padding! It's not super important, but if you feel like packing these bools into the spare bits up in `Decl` in a follow-on PR, you'd have my micro-optimizing thanks

File: include/swift/AST/Decl.h [hunk 21]
```diff
@@ -4517,6 +4521,8 @@ class AbstractFunctionDecl : public ValueDecl, public DeclContext {
 
   ImportAsMemberStatus IAMStatus;
 
+  BodyPostProcessor PostProcessor = nullptr;
```
Comment (DougGregor) [inline include/swift/AST/Decl.h]: Also totally fine going into a subsequent PR, but could we use a bit (up in `AbstractFunctionDeclBitfields`) to indicate the presence of a `BodyPostProcessor` and then put the actual post-processors into a `DenseMap` on `ASTContext`? It's a shame to waste a word in every function for these rare cases.

File: include/swift/AST/Module.h [hunk 22]
```diff
@@ -831,13 +831,13 @@ class SourceFile final : public FileUnit {
...
   /// The list of local type declarations in the source file.
-  TinyPtrVector<TypeDecl*> LocalTypeDecls;
+  llvm::SetVector<TypeDecl *> LocalTypeDecls;
 
   /// A set of special declaration attributes which require the
   /// Foundation module to be imported to work. If the foundation
   /// module is still not imported by the time type checking is
   /// complete, we diagnose.
-  std::map<DeclAttrKind, const DeclAttribute *> AttrsRequiringFoundation;
+  SmallPtrSet<const DeclAttribute *, 4> AttrsRequiringFoundation;
```
Comment (DougGregor) [inline include/swift/AST/Module.h]: Technically, this change will make the ordering of diagnostics based on this set non-deterministic.

**Review by DougGregor — APPROVED** [review]
Very nicely done! The post-processing does seem like the best approach within the scheme we have.

**Comment by CodaFi** [comment]
I was actually thinking of splitting this into its own file and moving it back into Parse.  From there I can expose a new phase for General Condition Resolution that will skip functions without bodies.  Then Parse can call the same hook and allow the methods that resolve function bodies for code completion to request condition clause evaluation should they need it.  I feel really bad about adding more bits here and this way I can revert all the recently-introduced nastiness.

File: include/swift/AST/Module.h (hunk 22 above)
Comment (CodaFi) [inline include/swift/AST/Module.h]: I'll make this a `SetVector`.

**Comment by CodaFi** [comment]
[Here's](https://github.com/apple/swift/compare/master...CodaFi:fine-imported-services#diff-15e077a6471dc4028bb78babfe1fc5b8R1) roughly what that version of the patch would look like.

**Comment by DougGregor** [comment]
I do like the layering improvements from keeping this in parse.

**Comment by CodaFi** [comment]
@jrose-apple Was this the [service-oriented approach](https://github.com/apple/swift/pull/5778/files#diff-d66eb1cd5371f343e115b73faf09d339R132) you had in mind?  

@swift-ci please test.

**Comment by CodaFi** [comment]
@swift-ci please smoke test.

File: lib/Sema/TypeChecker.cpp [hunk 23]
```diff
@@ -558,6 +558,12 @@ void swift::performTypeChecking(SourceFile &SF, TopLevelContext &TLC,
   if (SF.ASTStage == SourceFile::TypeChecked)
     return;
 
+  // Resolve condition clauses so name binding can accurately track imports.
+  {
+    SharedTimer timer("Condition resolution");
+    performConditionResolution(SF);
```
Comment (CodaFi) [inline lib/Sema/TypeChecker.cpp]: I wasn't sure if this should be folded in with NameBinding, but I err'd on the side of caution because of the concerns about slowdown for this feature.

**Comment by jrose-apple** [comment]
Hm. I didn't think about tying it to delaying a whole function body, but that might be good enough…although we still need to actually *do* that during normal compilation. (IIRC we currently only do it for code completion.)

I don't like the idea of Parse being able to import modules, but the benefit of having parseDelayedDecl Just Work might outweigh that.

Please do add some test cases that show `#if canImport(…)` in function bodies, in top-level contexts under -parse-as-library, and in closure bodies used to initialize a variable.

**Comment by CodaFi** [comment]
@jrose-apple How do [the tests](https://github.com/apple/swift/pull/5778/files#diff-4e6fc263adb1dba87f186983e7edc67f) look to you now?

File: test/Parse/ConditionalCompilation/can_import.swift [hunk 24]
```diff
@@ -0,0 +1,60 @@
...
+  public var boolValue: Bool {
+#if canImport(AppKit) || (canImport(UIKit) && (arch(i386) || arch(arm)))
+    return _value != 0
+#else
+    return _value
+#endif
+  }
+
+  /// Create an instance initialized to `value`.
+  @_transparent
```
Comment (jrose-apple) [inline test/Parse/ConditionalCompilation/can_import.swift]: Nitpick: please don't use `@_transparent` in tests where it's not relevant.

**Review by jrose-apple — COMMENTED** [review]
Thanks, but you're still missing

- `#if` at the top level of a -parse-as-library file.
- `#if` in the initializer of a property.
- `#if` in a non-primary file (which I forgot to ask for but which is worth testing)

**Comment by CodaFi** [comment]
OK, patched.

**Comment by jrose-apple** [comment]
Thanks! This is the property one I'm thinking of, sorry.

```swift
let value: Int = {
#if canImport(Swift)
  error() // expected-error {{blah blah blah}}
#else
  error() // should not happen
#endif
}()
```

File: test/Parse/ConditionalCompilation/can_import.swift [hunk 25]
```diff
@@ -0,0 +1,96 @@
+// RUN: %target-typecheck-verify-swift -parse-as-library
+// RUN: %target-typecheck-verify-swift -D WITH_PERFORM -primary-file %s %S/Inputs/can_import_nonprimary_file.swift
+
+public struct LibraryDependentBool : ExpressibleByBooleanLiteral {
+#if canImport(Swift)
+  var _description: String
+#endif
+
+  public let magicConstant: Int = {
```
Comment (CodaFi) [inline test/Parse/ConditionalCompilation/can_import.swift]: @jrose-apple Is this what you had in mind?

Comment (jrose-apple) [inline test/Parse/ConditionalCompilation/can_import.swift]: Yep, thanks!

**Comment by jrose-apple** [comment]
And since Doug already approved this I think that means it's good to go. :-)

@swift-ci Please test

**Comment by CodaFi** [comment]
@swift-ci please smoke test.

**Comment by CodaFi** [comment]
I'm going to give it a soft run to get the cobwebs out then a full test and merge to be safe.  

Thanks!

**Comment by CodaFi** [comment]
Full test happening [here for OS X](https://ci.swift.org/view/Pull%20Request/job/swift-PR-osx/4604/)

**Comment by CodaFi** [comment]
Thanks for sticking through.  I'm going to prepare the optimization patch for Doug and update the CHANGELOG in future PRs.

⛵️
